fix(input[date]): support years with more than 4 digits

LeeAdcock opened the pull request:
Change to regular expressions used for matching ISO date strings. Increases the number of digits allowed in a year from exactly four to instead be four or more. Also matches on the entire string, resolving ambiguity around extra characters at the start of the year, or after the timezone.

Closes https://github.com/angular/angular.js/issues/13735

LeeAdcock committed:
fix(input[date]): support years with more than 4 digits

Change to regular expressions used to match date formats that increases the number of digits
in a year. Previously exactly four digits were required, change allows four or more. Also
matches complete string, resolving ambiguity around extra characters at the beginning or
end.

LeeAdcock committed:
test(input[date]): improve test descriptions, removed erroneous console.log

LeeAdcock committed:
fix(input[date]): resolving jshint finding

LeeAdcock committed:
test(input[date]): removing unnecessary space, per jscs

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -657,6 +657,18 @@ describe('input', function() {
+      expect(+$rootScope.value).toBe(Date.UTC(10123, 2, 1, 0, 0, 0));
+
+      $rootScope.$apply(function() {
+        $rootScope.value = new Date(Date.UTC(10123, 2, 1, 0, 0, 0));
```
You should use a different value.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -856,6 +868,17 @@ describe('input', function() {
+      expect(+$rootScope.value).toBe(Date.UTC(10123, 0, 21));
+
+      $rootScope.$apply(function() {
+        $rootScope.value = new Date(Date.UTC(10321, 0, 21));
```
You should use a different value.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -1734,6 +1757,19 @@ describe('input', function() {
+        expect(+$rootScope.value).toBe(Date.UTC(10123, 0, 1, 0, 0, 0));
+
+        $rootScope.$apply(function() {
+          $rootScope.value = new Date(Date.UTC(10123, 0, 1, 0, 0, 0));
```
You should use a different value.

gkalpak commented:
There's no test for `datetime-local` :smiley:

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -1941,6 +1977,111 @@ describe('input', function() {
+        ['2010-01-01', false],                      // time must be specified
+        ['2010-01-0101:00:00.0000+01:01', false],   // missing date time seperator
+        ['2010-01-01V01:00:00.0000+01:01', false],  // invalid date time seperator
+        ['2010-01-01T01-00-00.0000+01:01', false],  // time must use period seperator
```
Actually, it must use colon separator.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -1941,6 +1977,111 @@ describe('input', function() {
+        ['2010-06-15T00:00:00.0000~01:01', false],  // timezone must have valid postive/negative indicator
+        ['2010-06-15T00:00:00.000001:01', false],   // timezone must have valid postive/negative indicator
+        ['2010-06-15T00:00:00.0000+00:01Z', false], // timezone invalid trailing characters
+        ['2010-06-15T00:00:00.0000+00:01 ', false], // timezone invalid trailing characters
```
Actually, according to the standard, the timezone offset can have one of three forms:

`+/-HH:mm`, `+/-HHmm`, `+/-HH`

(Feel free to add suppory for those as well :smiley:)

LeeAdcock committed:
fix(input[date]): iso date parsing, increase flexibility for timezone format

Support for parsing additional variations of ISO dates, specifically the minute offset in timezones
is now optional, as is the colon delimiter between the offset hour and minute.  Enhancements to the
tests were made to increase robustness and validate edge-case scenarios.

LeeAdcock commented:
Thank you for the feedback, I appreciate it.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -9,8 +9,8 @@
-// Regex code is obtained from SO: https://stackoverflow.com/questions/3143070/javascript-regex-iso-datetime#answer-3143231
-var ISO_DATE_REGEXP = /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)/;
+// Regex code was initially obtained from SO prior to modification: https://stackoverflow.com/questions/3143070/javascript-regex-iso-datetime#answer-3143231
+var ISO_DATE_REGEXP = /^\d{4,}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d(:?[0-5]\d)?|Z)$/;
```
Since we are not using the groups anywhere, it would be better for performance to make them non-capturing.
(As an added bonus, you'll have the chance to write a RegExp that includes `?::?` - not somethng you do everyday :grin:)

gkalpak replied:
BTW, I know I proposed this change, but there is currently an issue with IE/Edge (they can't parse a timezone offset when there is no colon), so additional steps might be needed. That said, I'm not sure when/if `ISO_DATE_REGEXP` is used after all, so let me look into it some more :)

LeeAdcock replied:
Likely to be a problem if the input model is populated directly with an ISO formatted text string, instead of from the browser native picker.

LeeAdcock replied:
I took a look at Edge in Saucelabs and found the same result when using the Date constructor for parsing both '2010-06-15T00:00:00.0000+**01:30**' as well as '2010-06-15T00:00:00.0000+**0130**'.

Is there a bug report for the Edge issue you referenced?

gkalpak replied:
I don't know if there is a bug report - it just came up a few days ago and I added a workaround to fix a similar issue in #13887.

gkalpak replied:
> Likely to be a problem if the input model is populated directly with an ISO formatted text string

The `ISO_DATE_REGEXP` is only used in an `ngModelController` parser, meaning it will only be applied when parsing the `$viewValue` to transform it to a `$modelValue`. The only case where the `$viewValue` can be an ISO-8601 date string, is on browsers that do not natively support `input[date]` and the users enter it by hand (which is a rather unlikely interaction imo).

Anyway, assuming we still need `ISO_DATE_REGEXP`, what needs to be done in order to properly support all valid timezone offsets on all supported browsers, is "fixing up" the date string before passing it to `new Date()`. That would include adding a colon if missing and add `00` minutes if omitted.

Thinking about it again, this is totally unrelated to the initial intent of this PR, so it's not a good idea to mix them. I suggest, leaving only changes to add support for >4-digit years in this PR and if you fancy feel free to submit another one with the required changes to properly support all valid timezone offset formats in another PR.

Sorry, I had you make those changes and now telling they don't belong here :disappointed:

LeeAdcock committed:
perf(input[date]): iso date matching regex should be non-capturing

LeeAdcock committed:
fix(input[date]): iso date parsing, Edge timezone offset requires colon and minutes

Narretz commented:
@gkalpak Can you take another look at this?

gkalpak commented:
@Narretz, will do. Thx for the heads-up ! (I hadn't notice the updates.)

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -1941,6 +1989,125 @@ describe('input', function() {
+        expect(ISO_DATE_REGEXP.test(date)).toBe(valid);
+      });
+
+      it('should be non-capturing', function() {
```
This test is not needed. Capturing vs Non-capturing is just an implementation detail.

LeeAdcock replied:
Happy to make this change, but would point out that if having the regex be non-capturing is important for performance, we should have a test that validates it. The regex was originally capturing, but that mistake wasn't noticed until this unit test was written. It's not unlikely that the next person that modifies it after us will unintentionally make it capturing again, unless there is a test enforces this style of implementation for performance gains.

gkalpak replied:
This is still an implementation detail. E.g. if we decide to add support for all timezone formats, this will change: we'll need some capturing groups. Unless it is identified as a performance issue (which it won't - it's more of a theoretical performance gain, unlikely to be noticable), it remains an implementation detail.

So, while it's nice to avoid capturing groups if you don't need them, having a test for it will only make it more difficult to add functionality later. Unit tests should assert the functionality/behavior, not the performance characteristics, imo.

gkalpak commented:
LGTM (expect for one minor comment).
Thx for working on this one, @LeeAdcock :+1:

LeeAdcock committed:
test(input[date]): removing test of non-functional regex implementation

LeeAdcock commented:
Thanks for the feedback. The implementation test has been removed.

gkalpak commented:
Backported to `v1.5.x` as d76951f.
